iminuit/_libiminuit.pyx

```diff
@@ -1424,3 +1590,81 @@ cdef class Minuit:
                 raise RuntimeError(
                     ('Cannot understand keyword %s. May be a typo?\n'
                      'The parameters are %r') % (k, parameters))
+
+    IF HAVE_NUMPY:
```

Comment:
Is it a good idea to put this `IF HAVE_NUMPY` before the `def np_matrix`?
I think it will just lead to users getting a cryptic `AttributeError` instead of the normal `ImportError` when they run a script that call `np_matrix`, no?
Suggest to change this back.

Btw, why is it `IF` and not `if`?

Reply:
IF is a compile-time if, not a run-time if. You can look it up in the Cython documentation. The methods np_* are now only added to the Minuit object when numpy was available during the compilation of the module.

That these methods were there even if numpy as not available was a bug, now it is consistent.

Reply:
All project I know use the "delayed import" in the function / method for optional dependencies.
Not the technique with the compile-time IF you introduce here, to hide the methods if the dependency is not available.
But ImportError is better than AttributeError if Numpy is not available, and also to "see" that the method exists even if no Numpy is installed, so that I can just install it, is the better user experience, no?

Reply:
The `np_*` methods should not be there in the first place when numpy is not available on your system. If they are there, this is an interface bug. They should not show up when you do `help(iminuit.Minuit)` in the interpreter.

The other projects you mention do it this way, because they don't know how to do a better job. I spend some hours on how to do this properly. The IF statement seems to be a recent addition to Cython. It probably was not available when these other projects were written.

Reply:
I also challenge the statement that ImportError is better than AttributeError. What is the error you should naively expect when you try to call a method that logically should not be there? You should get an AttributeError.

Reply:
Come one, that's just a matter of preference. There is no right or wrong here. Most other projects do it the way we had it and that I prefer. It's technically trivial to do it the way you do it for any Python project.

But that's a minor change here. Maybe we table the discussion or wait for someone else to chime in. And hopefully someone will have time / expertise to review the rest of the PR.

Reply:
No, it is not a matter of preference. In design, there is right and wrong. Once we agree on the criteria by which we judge a design, we can draw an objective conclusion. The subjectivity is  in the criteria by which to judge a design.

In the case of Python, the design criteria have been informally written down in form the Zen of Python. I like Python because of these criteria, for me they are intuitively obvious. Others have to accept them as authority.
"Beautiful is better than ugly."
"Explicit is better than implicit."
"Simple is better than complex."
"Readability counts."

In my own words: things should be obvious in Python, no false leads. If your class has a method that consistently fails whenever you call it, that method should not be there in the first place. It should not appear in the documentation, when you use the online help. Announcing methods in the online help that do not work is bad design. It is ugly, it is a false lead, the error is implicit, and it reduces readability of the documentation, because it is a distraction.

Reply:
Yes, let's talk in person about these things.

Still, here is the thing that I think you are not seeing yet. In the new version of iminuit, you have to decide at compile-time whether Numpy support is enabled or not. The new default will be Numpy Support ON, with Numpy Support OFF as an option. This is an acceptable increase in dependencies, because everyone uses Numpy anyway. Once Numpy Support is a compile-time decision and you turn it off, it makes no sense to have methods that require numpy to be present. It is more consistent to remove these methods then, because they are useless clutter.

Imagine you had a door with two door handles, one red and one blue. The door would only operate with the red handle, while the blue one would give you a mild electric shock. No designer would keep the blue handle around.